src/clients/patreon/patreonAgent.ts

```diff
@@ -23,195 +49,161 @@ export class PatreonAgent {
     existingPatreon: any,
-    patronDetails: any,
+    patronFullDetails: PatronFullDetails,
     usernameLower: string,
-    tokens: PatreonUserTokens,
-  ): Promise<PatreonDetails> {
-    const amountCents =
-      patronDetails.patreonMemberDetails.currently_entitled_amount_cents;
-    const lastChargeDate = new Date(
-      patronDetails.patreonMemberDetails.last_charge_date,
-    );
-
-    // Check payment received
+    patreonUserTokens: PatreonUserTokens,
+  ): Promise<PatronPublicDetails> {
+    // Check payment received to update currentPledgeExpiryDate
     const thirtyDaysAgo = new Date(Date.now() - 30 * 24 * 60 * 60 * 1000);
     const hasPaid =
-      patronDetails.patreonMemberDetails.last_charge_status &&
-      patronDetails.patreonMemberDetails.last_charge_status === 'Paid' &&
-      lastChargeDate &&
-      lastChargeDate > thirtyDaysAgo;
+      patronFullDetails.patronMemberDetails.lastChargeStatus &&
+      patronFullDetails.patronMemberDetails.lastChargeStatus === 'Paid' &&
+      patronFullDetails.patronMemberDetails.lastChargeDate &&
+      patronFullDetails.patronMemberDetails.lastChargeDate > thirtyDaysAgo;
     const currentPledgeExpiryDate = hasPaid
-      ? patronDetails.patreonMemberDetails.next_charge_date
+      ? patronFullDetails.patronMemberDetails.nextChargeDate
       : null;
 
-    const patreonUpdateDetails = {
-      patreonUserId: patronDetails.patreonUserId,
-      patreonUsersName: patronDetails.patreonUsersName,
+    const patreonRecordUpdateDetails = {
+      patreonUserId: patronFullDetails.patreonUserId,
       proavalonUsernameLower: usernameLower,
-      userAccessToken: tokens.userAccessToken,
-      userRefreshToken: tokens.userRefreshToken,
-      userAccessTokenExpiry: tokens.userAccessTokenExpiry,
-      amountCents,
-      // TODO-kev: Check this is accurate
+      userAccessToken: patreonUserTokens.userAccessToken,
+      userRefreshToken: patreonUserTokens.userRefreshToken,
+      userAccessTokenExpiry: patreonUserTokens.userAccessTokenExpiry,
+      amountCents:
+        patronFullDetails.patronMemberDetails.currentlyEntitledAmountCents,
       currentPledgeExpiryDate: currentPledgeExpiryDate,
     };
 
     if (existingPatreon) {
-      await patreonId.findOneAndUpdate(
+      // Due to limited testing capabilities with Patreon API return results:
+      // If currentPledgeExpiryDate is earlier than previously set expiry, then do not change
+      if (
+        existingPatreon.currentPledgeExpiryDate >
+        patreonRecordUpdateDetails.currentPledgeExpiryDate
+      ) {
+        patreonRecordUpdateDetails.currentPledgeExpiryDate =
+          existingPatreon.currentPledgeExpiryDate;
+      }
+
+      await patreonRecord.findOneAndUpdate(
         {
           proavalonUsernameLower: usernameLower,
         },
-        patreonUpdateDetails,
+        patreonRecordUpdateDetails,
       );
-    } else {
-      await patreonId.create(patreonUpdateDetails);
+    } else if (hasPaid) {
```

Comment:
```suggestion
    } else {
```
?

If this should be `else if (hasPaid)` then we should add an else and handle that case too.

Reply:
Added a comment to explain instead, but we are not reating a new patreonRecord for unpaid users. Hence else will do nothing